WP list edit feature

furinvader opened the pull request:
### Description

This PR introduces inplace editing for work packages in the list view.
Supersedes #4153 

After trying `angular-xeditable` and fiddling around with the current inplace edit implementation, I've found that implementing a slim new module was the quickest solution.

Be aware, that currently the **list view does not work** as it's supposed to.
### TODO
- [x] Enable subject field editing
- [x] Enable drop down field editing (Type, Priority, ...)
- [x] Save the work package
- [x] @ulferts Use schema properties in edit form templates (linke `maxlength`, `minlength` and so on) [TODO](https://github.com/furinvader/openproject/blob/experimental/new_wp_table/1/frontend/app/components/wp-edit/wp-edit-field.directive.html#L8)
  - furinvader/openproject#8
- [x] @ulferts, @oliverguenther Automatically remove non writable properties from the work package, when using `save` [TODO](https://github.com/furinvader/openproject/blob/experimental/new_wp_table/1/frontend/app/components/api/api-work-packages/api-work-packages.config.ts#L73)
  controller
  - furinvader/openproject#7
- [x] @ulferts prevent calls to `work_packages/:id/:attribute_name` (e.g. http://localhost:3000/api/v3/work_packages/144/assignee) if the value is not set for a work package
  - furinvader/openproject#10
- [x] @ulferts Reenable the context menu on the work packages table
  - furinvader/openproject#9
  - furinvader/openproject#14
- [x] Fix karma, protractor, specs and cukes
  - furinvader/openproject#11
  - furinvader/openproject#12
- [x] Remove work package querying from experimental
  - [x] PaginationService, etc.
  - [x] Grouping of fields
  - https://github.com/furinvader/openproject/pull/16
- [x] Handle inserting / removing columns, since new columns are empty.
  - https://github.com/furinvader/openproject/pull/16
- [x] Add possibility for error handling to the `wp-edit.directive`
- [x] @ulferts remove empty rows from table
  - furinvader/openproject#13
- [x] `wp-td.directive.ts` -> Show values depending on their type as defined by the schema
  - furinvader/openproject#18
- [x] Display sums (grouping and total) according to the schema 
  - furinvader/openproject#18
- [x] assignee (and possibly responsible) is considered a text field when it should be a drop down
- [x] after save, the altered value is not displayed in the table (although the field is updated)
- [x] empty fields (e.g. assignee not set) can not be edited. This is due to the field having a height of `0px` when the value is not set. 
- [x] if the user lacks the permission to edit a work package (e.g. he is not logged in), there are errors in the console.
- [x] Attributes that are not required and that have a `select` as their input element should have a blank value ("-") as their first value.
  - [x] Delete attributes, e.g. assignee
- [x] The assignee attribute does not get it's current value selected when entering inline edit mode.
- [x] Modifying filters in saved queries will lead to an error in the experimental API.
- [x] After saving an inplace-edit once, subsequent edits will succeed, but no longer update the table cell.

`...`
#### Missing tests
- [x] `halTransform` service
- [ ] `apiWorkPackages` service
- [ ] `apiWorkPackages` config - test the added convenience methods

`...`
#### Miscellaneous / Ideas
- [ ] Update type definitions file - best done in development. This item should be checked when everything else is finished.
- ~~Make WP schema accessible (especially `allowedValues`) via the linked resource 'schema'. 
  Currently, we have to request the WP edit form first, which is quite unnecessary.~~
- ~~Make query filters work again. Implement the `apiMetaData` service, where the meta result of the experimental API is handled. There we will map the old properties to the new ones (column names, custom fields etc.)~~
#### Out of Scope
- Use the new `wpEdit` module in the details view, replacing `inplace-edit`.
- Update work package table after submit using the new values
- Enable two-way binding between list and details view
- Handle custom fields
- Classes of the [`wp-edit-field.module.ts`](https://github.com/furinvader/openproject/blob/experimental/new_wp_table/1/frontend/app/components/wp-edit/wp-edit-field.module.ts#L65-L86), that should be added, if they introduce any new behaviour.
- The updated on cell is not updated after editing a work package.
- Work packages groups are not updated after changing an attribute.
- Changing between list view, split view and fullscreen view.
- When clicking in a text field the cursor should be at the position where the user clicked.
- Misleading animation after changing the sorting criteria (there are some seconds waiting between the animation and the actual change)
##### Documentation
- [ ] Add documentation for the API services
- [ ] Add `jsdoc`/`ngdoc` and 'enforce' its usage
#### Addresses WPs
- https://community.openproject.org/work_packages/22816
- https://community.openproject.org/work_packages/22551/activity
- https://community.openproject.org/work_packages/22549/activity
- https://community.openproject.org/work_packages/22552/activity
- https://community.openproject.org/work_packages/22055/activity

furinvader committed:
Refactor apiPaths service

furinvader committed:
Install restangular

furinvader committed:
Add apiV3 service based on restangular

furinvader committed:
Add apiExperimental service based on restangular

furinvader committed:
Add OpenProject type definitions

furinvader committed:
Add angular-xeditable package

furinvader committed:
Uninstall angular-xeditable

furinvader committed:
Add momentJS type definitions

furinvader committed:
Implement HAL transformation for restangular resources

furinvader committed:
Add restangular config

Enable caching for requests.

furinvader committed:
Don't transform non-resource elements

furinvader committed:
Fix source property for linked and embedded resources

furinvader committed:
Add data() method for HAL resources

It returns the plain object with the _links property.

furinvader committed:
Add type definitions

furinvader committed:
Add work packages API service

furinvader committed:
Add meta data service

Draft, WIP

furinvader committed:
Use WP API service in list view

This is a WIP version.
We want some of the code there to be moved
to the meta API service, or other appropriate ones.

furinvader committed:
Implement first version of the WP edit module

Disable links in the WP td directive as to
prevent opening the details view.
Also disable the table single-click handler

to prevent confusion. This actually might be
introduced back again, but can stay disabled
for development.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://furinvader-experimental-new-wp-table-1.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://furinvader-experimental-new-wp-table-1.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

furinvader commented:
Before we start collaborating, please wait until the referenced pull requests are merged and the branch is rebased.

ulferts commented on frontend/typings/restangular/restangular.d.ts:
```diff
@@ -0,0 +1,137 @@
+// Type definitions for Restangular v1.4.0
```
The restangular version in the bower.json is 1.5.1. So either only this comment is outdated or the type definitions have to be updated as well.

furinvader replied:
That is the content of a generated file, which we unfortunately have to commit.

furinvader committed:
Add lodash includes as an alias for contains

As we use an older version of lodash, _.includes is not a function.
So we have to add the alias ourselves.

This is related to the following restangular issue on github:
https://github.com/mgonto/restangular/issues/1314

furinvader committed:
Fix type definitions

furinvader committed:
Update hal transformer service

furinvader committed:
Add WP list params to WP api service

furinvader committed:
Rename the select field's allowedValues property to options

furinvader committed:
Rename collection method from 'all' to 'list'

furinvader committed:
Implement a wpEditField service

oliverguenther committed:
Fix group by parameter passing

oliverguenther committed:
Remove work package query execution in experimental

ulferts committed:
add aria-required on inputs

This is based on the work_package schema. There is a dedicated directive
for this so that we can easily extend this to also have additional
  attributes (e.g. required, maxlength, ...).

  For now those restrictions have not been added because we have no way
  as of now to present them in the way we present other errors as well.

  Therefore, we are waiting for the api to return the errors for now.

ulferts committed:
reenable context menu

oliverguenther committed:
Merge API responses

Sums are required to be present at all times, so restore that
metainformation from the V3.

ulferts committed:
avoid request on null link

If a linked resource is not set (e.g. assignee is null) the api v3
displays it as:

{
  ...snip...
  "_links": {
    "assignee": {
      "href": null
    }
    ...snip...
  },
  ...snip...
}

For such cases, we want to avoid issuing a request to ... null.

oliverguenther committed:
Separate Param Mapping service

Required for queries

ulferts committed:
use proper interpolation for I18n

ulferts committed:
adapt specs to changed fullscreen view navigation

oliverguenther committed:
Use param service in query

Uses `ApiParamMappingService.constructor` which sounds wrong... ?

oliverguenther committed:
Disable legacy experimental calls

furinvader committed:
Replace emptyObjectPromise with $q.when({})

furinvader committed:
Test current behaviour of the HAL transformation

furinvader commented:
Sorry, I have rebased the branch and force pushed it :/

furinvader committed:
Remove redundant restangularize method

furinvader committed:
Add test for null href return value of hal links

furinvader committed:
Automatically set linked properties on object transformation

furinvader committed:
Remove setProperty and setProperties

Those methods are replaced by plainly using links.

furinvader committed:
Refactor transformation

Remove the useless Restangular methods from the element
and make the added properties enumerable.

furinvader committed:
Prepend special properties with a $

furinvader committed:
Get rid of the _source property

furinvader committed:
Refactor transformation service

... for better readability.

ulferts committed:
adapt spec to altered json

furinvader committed:
Add test for non-transformed objects

ulferts committed:
adapt js configuration to changed link

ulferts committed:
adapt cuke to altered navigation

furinvader committed:
Implement $plain method

ulferts committed:
extend focus js to all focusable elements

furinvader committed:
Change the transformation service to a factory

To keep the restangular context from which the element
was transformed, so the other configurations and
transformation are applied to the results.

furinvader committed:
Move hal transform service

furinvader committed:
Move apiV3 config into run block

furinvader committed:
Implement halTransform service

furinvader committed:
Make the halTransform service a function

furinvader committed:
Change implementation of HAL resources

The elements transformed by Restangular
are now actual objects of the HalResource type.

The Restangular element is saved as the $source
property of the resource.

We now can implement classes for each type of
resource returned by the API.

furinvader committed:
Fix properties for nested embedded resources

Previously, the embedded resource would not have
the properties of its source.

furinvader committed:
Fix work package save method

furinvader commented:
After implementing the transformation service using actual Angular services that provide the necessary classes, I should do the same for the WP edit field classes.
The current implementation is pretty dirty with the `$injector` service being set as a static property for every field subclass.

ulferts committed:
wait for the table to be loaded before proceeding with the test

ulferts committed:
introduce collectionResource

It is then employed to then get the elements of the embedded list to populate the work packages table

ulferts committed:
adapt context menu to changes in halResource

ulferts committed:
add 'title' to list of properties used for rendering wp table td

furinvader committed:
Merge pull request #15 from ulferts/experimental/new_wp_table/1_fix_wp_table_td_content

add 'title' to list of properties used for rendering wp table td

furinvader committed:
Merge pull request #14 from ulferts/experimental/new_wp_table/1_adapt_context_menu

adapt context menu to changes in halResource

furinvader committed:
Merge pull request #13 from ulferts/experimental/new_wp_table/1_collection_resource

introduce collectionResource

furinvader committed:
Use Array.isArray when transforming an array resource

furinvader committed:
Transform properties of embedded resources

ulferts committed:
fix context menu karma tests

furinvader committed:
Implement HalLink service

furinvader committed:
Refactor property transformation of embedded resources

furinvader committed:
Refactor property transformation in general

ulferts committed:
fix apiPath tests that got broken by memoization

ulferts committed:
adapt wp.td directive test

ulferts committed:
no longer call experimental api on column adding

the api v3 already delivers all the information we need

ulferts committed:
wip - merge artifact on wpCollection

furinvader committed:
Remove $fetchCollection

Collections are now independent from
Restangular specific behaviour.

furinvader committed:
Implement $toFunc method for HalLink

The method is a wrapper for the $fetch method
and can now be used to "hide" the actual link object
from the user.

furinvader committed:
Add test for query params

furinvader committed:
Implement a HalLink factory method using $toFunc

furinvader committed:
Remove redundant test

furinvader committed:
Stop setting linked resources as properties

furinvader committed:
Make links of the resource callable again

ulferts committed:
Merge pull request #12 from ulferts/experimental/new_wp_table/1_fix_tests

Experimental/new wp table/1 fix tests

ulferts committed:
use consistent naming for columns/filters across apiv3 and experimental

ulferts committed:
remove unused imports

ulferts committed:
differntiate work_package path called based on existence of project information

ulferts committed:
correclty call v3 with filters and sortBy

ulferts committed:
remove no longer valid parameter

ulferts committed:
remove no longer necessary mapping service

furinvader committed:
Make linked and embedded resources accessible via properties

ulferts committed:
move v3 naming into own module

ulferts committed:
remove no longer applicable tests

furinvader committed:
Remove premature doc comments

furinvader committed:
Provide a name attribute that uses the self link's title

This is useful when the resource is not loaded, but the
name attribute is needed. Usually the name attribute is
the title attribute of the link.

furinvader committed:
Move the wp-edit-field files in their own directory

furinvader committed:
Fix select field options

furinvader committed:
Rename $halTransformed to $isHal

furinvader committed:
Refactor hal tests

furinvader committed:
Merge branch 'experimental/new_wp_table/1' of github.com:furinvader/openproject into experimental/new_wp_table/1

ulferts committed:
fix pagination

ulferts committed:
fix pagination

furinvader committed:
Add $plain property to Restangular results

ulferts committed:
rename remaining wp-column to wp-td

ulferts committed:
fix grouping

furinvader committed:
Remove unnecessary promise handling

oliverguenther committed:
Merge pull request #16 from ulferts/experimental/new_wp_table/remove_wp_experimental

Experimental/new wp table/remove wp experimental

furinvader committed:
Remove angular.copy from HalResource tests

ulferts committed:
add sumsSchema to api

furinvader committed:
Make arrays of linked resources work

furinvader committed:
Set loaded schema resources as properties of the work package

furinvader committed:
Fix WP save method for linked resources

furinvader committed:
Show generic error message on WP update failure

furinvader committed:
Merge branch 'experimental/new_wp_table/1' of github.com:furinvader/openproject into experimental/new_wp_table/1

oliverguenther committed:
patch the work package by sending the form's payload

* Get the form
* Get the schema (part of the form)
* Get the payload (part of the form)
* add all attributes of the work package to the payload
* send the payload

furinvader committed:
Add QueryParams interface

ulferts committed:
reenable wp row selection

furinvader committed:
Cache the form of WP getForm()

ulferts committed:
Merge pull request #17 from ulferts/experimental/new_wp_table/reenable_row_selection

reenable wp row selection

furinvader committed:
Remove angular.copy call to enhance performance

ulferts committed:
consider a resource loaded only after it is loaded

ulferts committed:
indentation

ulferts committed:
use the halResource to display the table sums

ulferts committed:
add specific formatting

furinvader committed:
Merge branch 'experimental/new_wp_table/1' of github.com:furinvader/openproject into experimental/new_wp_table/1

ulferts committed:
completely fix showing sums

ulferts committed:
reintoduce special handling of percentageDone field

ulferts committed:
remove no longer necessary variables from signature

oliverguenther committed:
Use errors from API

ulferts committed:
more camelcase for the experimental api

ulferts committed:
remove unused method

furinvader committed:
Prevent selection click when editing field

ulferts committed:
Merge branch 'dev' into experimental/new_wp_table/1

ulferts committed:
Merge pull request #18 from ulferts/experimental/new_wp_table/fix_sums

Experimental/new wp table/fix sums

oliverguenther committed:
Control single promise in wp-resource#save

ulferts commented:
https://community.openproject.org/work_packages/22816

oliverguenther committed:
Restore default resource check

oliverguenther committed:
Merge pull request #7 from ulferts/experimental/new_wp_table/1_saving

Experimental/new wp table/1 saving

ulferts committed:
remove unused methods/classes

ulferts committed:
adapt spec to no longer expect the view to have work_packages

ulferts committed:
fix default sorting by parent

ulferts committed:
adapt specs to altered implementation

ulferts committed:
fix child indentation on wp table

ulferts committed:
fix default query initialization again

oliverguenther committed:
Fix all karma tests exepect group sums

ulferts committed:
adapt sum specs

* as we are no longer needing total-sums.directive it is deleted

oliverguenther committed:
Add separate experimental and V3 mocks

This should fix protractor tests in the process.

ulferts committed:
correctly handle inline edit for attributes of type "User"

ulferts committed:
prevent trying to get into edit mode when user is not allowed to edit

ulferts committed:
remove unnecessary code

There will never be a params[:format] on this controller so we don't need a case statement for it

ulferts commented:
* https://community.openproject.org/work_packages/22551/activity
* https://community.openproject.org/work_packages/22549/activity
* https://community.openproject.org/work_packages/22552/activity

oliverguenther committed:
Watch for attribute changes in the resource

ulferts commented:
https://community.openproject.org/work_packages/22055/activity

ulferts committed:
use camelCase for queryId

oliverguenther committed:
Show default placeholder when no value in td

oliverguenther committed:
Track objects by href and set self link

This adds the self href attribute to a HALResource whenever the object
has a self link.

This is used for tracking, e.g., selected objects.

ulferts committed:
limit which fields are editable

ulferts committed:
rename parameter in spec

ulferts committed:
adding version and category to lists of supported inline edit types

oliverguenther committed:
Track by 'null' object which is transformed before saving

The track by causes a 'null' href to land in the wp resource for changed links
which is not accepted by the API.

ulferts committed:
ensure the schema to be loaded witin a row
